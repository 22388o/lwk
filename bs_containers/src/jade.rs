use std::env;

use testcontainers::{core::WaitFor, Image, ImageArgs};

pub const EMULATOR_PORT: u16 = 30_121;

#[derive(Debug, Default)]
pub struct JadeEmulator;

#[derive(Clone, Debug, Default)]
struct Args;

impl ImageArgs for Args {
    fn into_iterator(self) -> Box<dyn Iterator<Item = String>> {
        let args = ["bash".to_string()];
        Box::new(args.into_iter())
    }
}

impl Image for JadeEmulator {
    type Args = ();

    fn name(&self) -> String {
        // TODO Change with blockstream official jade emulator
        env::var("JADE_EMULATOR_IMAGE_NAME").unwrap_or("xenoky/local-jade-emulator".into())
    }

    fn tag(&self) -> String {
        env::var("JADE_EMULATOR_IMAGE_VERSION").unwrap_or("latest".into())
    }

    fn ready_conditions(&self) -> Vec<WaitFor> {
        vec![WaitFor::StdOutMessage {
            message: "char device redirected".into(),
        }]
    }

    fn expose_ports(&self) -> Vec<u16> {
        [EMULATOR_PORT].into()
    }
}
