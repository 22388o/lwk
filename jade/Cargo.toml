[package]
name = "jade"
version = "0.1.0"
edition = "2021"

[dependencies]
# anyhow = "1"
ciborium = "0.2.1"
hex = "0.4.3"
rand = "0.8.5"
# reqwest = { version = "0.11.19", features = ["blocking", "json"] }
serde = "1"
serde_bytes = "0.11.12"
elements = { version = "0.23.0", features = ["base64", "serde"] }
serialport = { version = "4.2.1", optional = true }
thiserror = "1"

# FIXME: use minreq because it's the http client of the rust-bitcoin ecosystem
ureq = { version = "2.7.1", features = ["json"] }
elements-miniscript = { git = "https://github.com/ElementsProject/elements-miniscript.git", rev = "27b6e37af2cc850b20b5c4322b3673dcd452cd08" }

# tokio = { version = "1.32.0", features = ["rt", "macros"] }
# tracing = "0.1.37"
# tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }

[features]
serial = ["serialport"]

[[test]]
name = "e2e"
path = "test/e2e.rs"

[dev-dependencies]
insta = { version = "1.31.0", features = ["yaml"] }
tempfile = "3.8.0"
bs_containers = { path = "../bs_containers" }
base64 = "0.21.4"
serde_json = "1"
