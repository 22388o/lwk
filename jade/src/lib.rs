#![doc = include_str!("../README.md")]

use std::{
    collections::HashMap,
    io::ErrorKind,
    time::{SystemTime, UNIX_EPOCH},
};

use connection::Connection;
use elements::bitcoin::bip32::{DerivationPath, ExtendedPubKey, Fingerprint};
use get_receive_address::GetReceiveAddressParams;
use protocol::{
    AuthResult, AuthUserParams, DebugSetMnemonicParams, EntropyParams, EpochParams,
    GetSignatureParams, GetXpubParams, HandshakeData, HandshakeInitParams, IsAuthResult, Params,
    RegisteredMultisig, Request, Response, SignMessageParams, UpdatePinserverParams,
    VersionInfoResult,
};
use rand::RngCore;
use register_multisig::RegisterMultisigParams;
use serde::de::DeserializeOwned;
use serde_bytes::ByteBuf;
use sign_liquid_tx::{SignLiquidTxParams, TxInputParams};

pub mod connection;
pub mod consts;
pub mod error;
pub mod get_receive_address;
pub mod mutex_jade;
mod network;
pub mod protocol;
pub mod register_multisig;
pub mod sign_liquid_tx;
pub mod sign_pset;
pub mod unlock;

pub use consts::{BAUD_RATE, TIMEOUT};
pub use error::Error;
pub use network::Network;

#[cfg(feature = "serial")]
pub use serialport;

pub type Result<T> = std::result::Result<T, error::Error>;

#[derive(Debug)]
pub struct Jade {
    /// Jade working via emulator(tcp), physical(serial/bluetooth)
    conn: Connection,

    /// The network
    network: crate::Network,

    /// Cached master xpub
    master_xpub: Option<ExtendedPubKey>,
}

#[cfg(feature = "serial")]
#[derive(Debug)]
pub struct SerialJade {
    pub jade: Jade,
    pub network: Network,
    pub path: String,
    pub version: VersionInfoResult,
}

impl Jade {
    pub fn new(conn: Connection, network: Network) -> Self {
        Self {
            conn,
            network,
            master_xpub: None,
        }
    }

    #[cfg(feature = "serial")]
    pub fn scan_serial() -> Result<Option<SerialJade>> {
        let ports = serialport::available_ports()?;
        if ports.is_empty() {
            Ok(None)
        } else {
            // todo: loop through ports?
            let path = ports[0].port_name.clone();
            let port = serialport::new(&path, BAUD_RATE).timeout(TIMEOUT).open()?;

            let network = Network::TestnetLiquid; // todo: network selection
            let mut jade = Jade::new(port.into(), network);
            let version = jade.version_info()?;

            Ok(Some(SerialJade {
                jade,
                network,
                path,
                version,
            }))
        }
    }

    fn check_network(&self, passed: Network) -> Result<()> {
        let init = self.network;
        if passed != init {
            Err(Error::MismatchingXpub { init, passed })
        } else {
            Ok(())
        }
    }

    pub fn ping(&mut self) -> Result<u8> {
        self.send_request("ping", None)
    }

    pub fn logout(&mut self) -> Result<bool> {
        self.send_request("logout", None)
    }

    pub fn version_info(&mut self) -> Result<VersionInfoResult> {
        self.send_request("get_version_info", None)
    }

    pub fn set_epoch(&mut self, epoch: u64) -> Result<bool> {
        let params = Params::Epoch(EpochParams { epoch });
        self.send_request("set_epoch", Some(params))
    }

    pub fn add_entropy(&mut self, entropy: &[u8]) -> Result<bool> {
        let params = Params::Entropy(EntropyParams {
            entropy: entropy.to_vec(),
        });
        self.send_request("add_entropy", Some(params))
    }

    pub fn auth_user(&mut self) -> Result<IsAuthResult<String>> {
        let epoch = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(error::Error::SystemTimeError)?
            .as_secs();
        let params = Params::AuthUser(AuthUserParams {
            network: self.network,
            epoch,
        });
        self.send_request("auth_user", Some(params))
    }

    pub fn handshake_init(
        &mut self,
        params: HandshakeInitParams,
    ) -> Result<AuthResult<HandshakeData>> {
        let params = Params::HandshakeInit(params);
        self.send_request("handshake_init", Some(params))
    }

    pub fn update_pinserver(&mut self, params: UpdatePinserverParams) -> Result<bool> {
        let params = Params::UpdatePinServer(params);
        self.send_request("update_pinserver", Some(params))
    }

    pub fn handshake_complete(
        &mut self,
        params: protocol::HandshakeCompleteParams,
    ) -> Result<bool> {
        let params = Params::HandshakeComplete(params);
        self.send_request("handshake_complete", Some(params))
    }

    fn inner_get_xpub(&mut self, params: GetXpubParams) -> Result<ExtendedPubKey> {
        self.check_network(params.network)?;
        let params = Params::GetXpub(params);
        self.send_request("get_xpub", Some(params))
    }

    pub fn get_xpub(&mut self, params: GetXpubParams) -> Result<ExtendedPubKey> {
        if params.path.is_empty() {
            self.get_master_xpub()
        } else {
            self.inner_get_xpub(params)
        }
    }

    pub fn fingerprint(&mut self) -> Result<Fingerprint> {
        Ok(self.get_master_xpub()?.fingerprint())
    }

    pub fn get_master_xpub(&mut self) -> Result<ExtendedPubKey> {
        if self.master_xpub.is_none() {
            let master_xpub = self.inner_get_xpub(GetXpubParams {
                network: self.network,
                path: vec![],
            })?;
            self.master_xpub = Some(master_xpub);
        }
        Ok(self.master_xpub.expect("ensure it is some before"))
    }

    pub fn get_receive_address(&mut self, params: GetReceiveAddressParams) -> Result<String> {
        self.check_network(params.network)?;
        let params = Params::GetReceiveAddress(params);
        self.send_request("get_receive_address", Some(params))
    }

    pub fn sign_message(&mut self, params: SignMessageParams) -> Result<ByteBuf> {
        let params = Params::SignMessage(params);
        self.send_request("sign_message", Some(params))
    }

    pub fn get_signature_for_msg(&mut self, params: GetSignatureParams) -> Result<String> {
        let params = Params::GetSignature(params);
        self.send_request("get_signature", Some(params))
    }

    pub fn get_signature_for_tx(&mut self, params: GetSignatureParams) -> Result<ByteBuf> {
        let params = Params::GetSignature(params);
        self.send_request("get_signature", Some(params))
    }

    pub fn sign_liquid_tx(&mut self, params: SignLiquidTxParams) -> Result<bool> {
        self.check_network(params.network)?;
        let params = Params::SignLiquidTx(params);
        self.send_request("sign_liquid_tx", Some(params))
    }

    pub fn tx_input(&mut self, params: TxInputParams) -> Result<ByteBuf> {
        let params = Params::TxInput(params);
        self.send_request("tx_input", Some(params))
    }

    pub fn debug_set_mnemonic(&mut self, params: DebugSetMnemonicParams) -> Result<bool> {
        let params = Params::DebugSetMnemonic(params);
        self.send_request("debug_set_mnemonic", Some(params))
    }

    pub fn register_multisig(&mut self, params: RegisterMultisigParams) -> Result<bool> {
        self.check_network(params.network)?;
        let params = Params::RegisterMultisig(params);
        self.send_request("register_multisig", Some(params))
    }

    pub fn get_registered_multisigs(&mut self) -> Result<HashMap<String, RegisteredMultisig>> {
        self.send_request("get_registered_multisigs", None)
    }

    fn send_request<T>(&mut self, method: &str, params: Option<Params>) -> Result<T>
    where
        T: std::fmt::Debug + DeserializeOwned,
    {
        let mut rng = rand::thread_rng();
        let id = rng.next_u32().to_string();
        let req = Request {
            id,
            method: method.into(),
            params,
        };
        let mut buf = Vec::new();
        ciborium::into_writer(&req, &mut buf)?;
        tracing::debug!(
            "\n--->\t{:#?}\n\t({} bytes) {}",
            &req,
            buf.len(),
            &hex::encode(&buf),
        );

        self.conn.write_all(&buf).map_err(|e| Error::IoError(e))?; // not sure why the map_err is needed

        let mut rx = [0u8; 4096];

        let mut total = 0;
        loop {
            match self.conn.read(&mut rx[total..]) {
                Ok(len) => {
                    total += len;
                    match ciborium::from_reader::<Response<T>, &[u8]>(&rx[..total]) {
                        Ok(r) => {
                            if let Some(result) = r.result {
                                tracing::debug!(
                                    "\n<---\t{:?}\n\t({} bytes) {}",
                                    &result,
                                    total,
                                    hex::encode(&rx[..total])
                                );
                                return Ok(result);
                            }
                            if let Some(error) = r.error {
                                return Err(Error::JadeError(error));
                            }
                            return Err(Error::JadeNeitherErrorNorResult);
                        }

                        Err(e) => {
                            let res = ciborium::from_reader::<ciborium::Value, &[u8]>(&rx[..total]);
                            if let Ok(value) = res {
                                // The value returned is a valid CBOR, but our structs doesn't map it correctly
                                dbg!(&value);
                                return Err(Error::Des(e));
                            }

                            if len == 0 {
                                // There is no more data coming from jade and we can't parse its message, return error
                                return Err(Error::Des(e));
                            } else {
                                // it may be the parsing failed because there is other data to be read
                            }
                        }
                    }
                }
                Err(e) => {
                    if e.kind() != ErrorKind::Interrupted {
                        dbg!(&e);
                        return Err(Error::IoError(e));
                    }
                }
            }
        }
    }
}

pub fn derivation_path_to_vec(path: &DerivationPath) -> Vec<u32> {
    path.into_iter().map(|e| (*e).into()).collect()
}
