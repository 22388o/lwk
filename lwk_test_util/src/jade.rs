use crate::init_logging;
use elements::{
    bitcoin::PublicKey,
    hashes::{hex::FromHex, sha256, Hash},
    secp256k1_zkp::{ecdsa::Signature, Message, Secp256k1},
};
use lwk_containers::testcontainers::{clients::Cli, Container};
use lwk_containers::{JadeEmulator, PinServer, EMULATOR_PORT, PIN_SERVER_PORT};
use lwk_jade::{
    mutex_jade::MutexJade,
    protocol::{DebugSetMnemonicParams, HandshakeInitParams, UpdatePinserverParams},
    Jade, Network,
};
use tempfile::TempDir;

pub fn pin_server_verify(params: &HandshakeInitParams, pin_server_pub_key: &PublicKey) {
    let ske_bytes = Vec::<u8>::from_hex(&params.ske).unwrap();
    let ske_hash = sha256::Hash::hash(&ske_bytes);

    let signature_bytes = Vec::<u8>::from_hex(&params.sig).unwrap();
    let signature = Signature::from_compact(&signature_bytes).unwrap();

    let message = Message::from_digest_slice(&ske_hash[..]).unwrap();

    let verify = Secp256k1::verification_only()
        .verify_ecdsa(&message, &signature, &pin_server_pub_key.inner)
        .is_ok();
    assert!(verify);
}

/// A struct for Jade testing with emulator
pub struct TestJadeEmulator<'a> {
    pub jade: Jade,
    // Keep the containers and temp dir so they are not dropped.
    _jade_emul: Container<'a, JadeEmulator>,
    _pin_server: Option<Container<'a, PinServer>>,
    _pin_server_dir: Option<TempDir>,
}

impl<'a> TestJadeEmulator<'a> {
    /// Jade with emulator
    pub fn new(docker: &'a Cli) -> Self {
        init_logging();
        let container = docker.run(JadeEmulator);
        let port = container.get_host_port_ipv4(EMULATOR_PORT);
        let stream = std::net::TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
        let network = Network::LocaltestLiquid;
        let jade = Jade::new(stream.into(), network);
        Self {
            jade,
            _jade_emul: container,
            _pin_server: None,
            _pin_server_dir: None,
        }
    }

    /// Set a mnemonic
    pub fn set_debug_mnemonic(&mut self, mnemonic: &str) {
        let params = DebugSetMnemonicParams {
            mnemonic: mnemonic.to_string(),
            passphrase: None,
            temporary_wallet: false,
        };
        let result = self.jade.debug_set_mnemonic(params).unwrap();
        assert!(result);
    }

    /// Jade with emulator and dedicated pin server
    pub fn new_with_pin(docker: &'a Cli) -> Self {
        let mut test_jade_emul = Self::new(docker);

        let tempdir = PinServer::tempdir().unwrap();
        let pin_server = PinServer::new(&tempdir).unwrap();
        let pin_server_pub_key = *pin_server.pub_key();
        assert_eq!(pin_server_pub_key.to_bytes().len(), 33);
        let pin_container = docker.run(pin_server);
        let port = pin_container.get_host_port_ipv4(PIN_SERVER_PORT);
        let pin_server_url = format!("http://127.0.0.1:{}", port);

        let params = UpdatePinserverParams {
            reset_details: false,
            reset_certificate: false,
            url_a: pin_server_url.clone(),
            url_b: "".to_string(),
            pubkey: pin_server_pub_key.to_bytes(),
            certificate: "".into(),
        };

        let result = test_jade_emul.jade.update_pinserver(params).unwrap();
        assert!(result);

        test_jade_emul.jade.unlock().unwrap();

        test_jade_emul._pin_server = Some(pin_container);
        test_jade_emul._pin_server_dir = Some(tempdir);
        test_jade_emul
    }
}

/// A struct for MutexJade testing with emulator
pub struct TestMutexJadeEmulator<'a> {
    pub jade: MutexJade,
    // Keep the containers and temp dir so they are not dropped.
    _jade_emul: Container<'a, JadeEmulator>,
    _pin_server: Option<Container<'a, PinServer>>,
    _pin_server_dir: Option<TempDir>,
}

impl<'a> TestMutexJadeEmulator<'a> {
    /// MutexJade with emulator
    pub fn new(test_jade_emul: TestJadeEmulator<'a>) -> Self {
        let TestJadeEmulator {
            jade,
            _jade_emul,
            _pin_server,
            _pin_server_dir,
        } = test_jade_emul;
        let jade = MutexJade::new(jade);
        Self {
            jade,
            _jade_emul,
            _pin_server,
            _pin_server_dir,
        }
    }
}
