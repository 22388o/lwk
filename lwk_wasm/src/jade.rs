use crate::{
    serial::{get_jade_serial, WebSerial},
    Error, Network,
};
use lwk_jade::asyncr;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub struct Jade {
    inner: asyncr::Jade<WebSerial>,
    _port: web_sys::SerialPort,
}

#[wasm_bindgen]
impl Jade {
    /// Creates a Jade from Web Serial
    #[wasm_bindgen(constructor)]
    pub async fn from_serial(network: Network) -> Result<Jade, Error> {
        let port = get_jade_serial(false).await?;
        let web_serial = WebSerial::new(&port)?;

        let inner = asyncr::Jade::new(web_serial, network.into());
        Ok(Jade { inner, _port: port })
    }

    pub async fn get_version(&self) -> Result<JsValue, Error> {
        let version = self.inner.version_info().await?;
        Ok(serde_wasm_bindgen::to_value(&version)?)
    }
}
