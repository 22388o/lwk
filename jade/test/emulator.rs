use std::time::UNIX_EPOCH;

use jade::Jade;
use testcontainers::{clients, core::WaitFor, Image, ImageArgs};

const PORT: u16 = 30_121;

#[derive(Debug, Default)]
pub struct JadeEmulator;

#[derive(Clone, Debug, Default)]
pub struct Args;

impl ImageArgs for Args {
    fn into_iterator(self) -> Box<dyn Iterator<Item = String>> {
        let args = ["bash".to_string()];
        Box::new(args.into_iter())
    }
}

impl Image for JadeEmulator {
    type Args = ();

    fn name(&self) -> String {
        "xenoky/local-jade-emulator".into() // TODO Change with blockstream official jade emulator
    }

    fn tag(&self) -> String {
        "latest".into()
    }

    fn ready_conditions(&self) -> Vec<WaitFor> {
        vec![WaitFor::StdOutMessage {
            message: "char device redirected".into(),
        }]
    }

    fn expose_ports(&self) -> Vec<u16> {
        [PORT].into()
    }
}

const _TEST_MNEMONIC: &str = "fish inner face ginger orchard permit
                             useful method fence kidney chuckle party
                             favorite sunset draw limb science crane
                             oval letter slot invite sadness banana";

#[test]
fn entropy() {
    let docker = clients::Cli::default();
    let container = docker.run(JadeEmulator::default());
    let port = container.get_host_port_ipv4(PORT);
    let stream = std::net::TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
    let mut jade_api = Jade::new(stream.into());

    let result = jade_api.add_entropy(&[1, 2, 3, 4]).unwrap();
    insta::assert_yaml_snapshot!(result);
}

#[test]
fn epoch() {
    let docker = clients::Cli::default();
    let container = docker.run(JadeEmulator::default());
    let port = container.get_host_port_ipv4(PORT);
    let stream = std::net::TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
    let mut jade_api = Jade::new(stream.into());

    let seconds = std::time::SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let result = jade_api.set_epoch(seconds).unwrap();
    insta::assert_yaml_snapshot!(result);
}

#[test]
fn ping() {
    let docker = clients::Cli::default();
    let container = docker.run(JadeEmulator::default());
    let port = container.get_host_port_ipv4(PORT);
    let stream = std::net::TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
    let mut jade_api = Jade::new(stream.into());

    let result = jade_api.ping().unwrap();
    insta::assert_yaml_snapshot!(result);
}

#[test]
fn version() {
    let docker = clients::Cli::default();
    let container = docker.run(JadeEmulator::default());
    let port = container.get_host_port_ipv4(PORT);
    let stream = std::net::TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
    let mut jade_api = Jade::new(stream.into());

    let result = jade_api.version_info().unwrap();
    insta::assert_yaml_snapshot!(result);
}
