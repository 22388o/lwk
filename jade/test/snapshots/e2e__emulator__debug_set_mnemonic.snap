---
source: jade/test/emulator.rs
expression: result
---
JADE_VERSION: "1"
JADE_OTA_MAX_CHUNK: 4096
JADE_CONFIG: NORADIO
BOARD_TYPE: DEV
JADE_FEATURES: DEV
IDF_VERSION: v5.0.2
CHIP_FEATURES: "32000000"
EFUSEMAC: "000000000000"
BATTERY_STATUS: 0
JADE_STATE: READY
JADE_NETWORKS: ALL
JADE_HAS_PIN: false

